Devservice for `amqp` can't load channels on dev ui

### Describe the bug

When poking around dev ui for my `amqp` (`quarkus-smallrye-reactive-messaging-amqp`) enabled app, I found that when trying to access the "channels" ui section, nothing ends up loading. On the console output, I get:

```
2023-11-06 14:04:34,653 ERROR [io.qua.dev.run.jso.JsonRpcCodec] (executor-thread-1) Error in JsonRPC Call: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at io.smallrye.mutiny.unchecked.UncheckedSupplier.lambda$toSupplier$0(UncheckedSupplier.java:45)
	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.devui.runtime.comms.JsonRpcRouter.lambda$invoke$0(JsonRpcRouter.java:98)
	at io.smallrye.mutiny.unchecked.UncheckedSupplier.lambda$toSupplier$0(UncheckedSupplier.java:41)
	... 11 more
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.base/java.util.Objects.checkIndex(Objects.java:361)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at io.quarkus.smallrye.reactivemessaging.runtime.QuarkusMediatorConfiguration.getOutgoing(QuarkusMediatorConfiguration.java:122)
	at io.quarkus.smallrye.reactivemessaging.runtime.devui.DevReactiveMessagingInfos$1.get(DevReactiveMessagingInfos.java:61)
	at io.quarkus.smallrye.reactivemessaging.runtime.devui.DevReactiveMessagingInfos$1.get(DevReactiveMessagingInfos.java:28)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:32)
	at io.quarkus.smallrye.reactivemessaging.runtime.devui.DevReactiveMessagingInfos.getChannels(DevReactiveMessagingInfos.java:110)
	at io.quarkus.smallrye.reactivemessaging.runtime.devui.ReactiveMessagingJsonRpcService.getInfo(ReactiveMessagingJsonRpcService.java:16)
	at io.quarkus.smallrye.reactivemessaging.runtime.devui.ReactiveMessagingJsonRpcService_ClientProxy.getInfo(Unknown Source)
	... 17 more
```

### Expected behavior

The devui should load channels correctly.

### Actual behavior

An indexoutofbounds exception occurs

### How to Reproduce?

Reproducer: https://github.com/GregJohnStewart/calc-quarkus

 1. Start service `./gradlew quarkusDev`
 2. Navigate to dev ui `localhost:8080/'
 3. Click on "Channels" under "Smallrye Reactive Messaging": 
     ![image](https://github.com/quarkusio/quarkus/assets/7083701/2fd2be1d-2f52-4f40-8478-e386a2a44797)


### Output of `uname -a` or `ver`

Linux dev.local 6.5.9-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Oct 25 20:40:49 UTC 2023 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.8" 2023-07-18 OpenJDK Runtime Environment (Red_Hat-17.0.8.0.7-1.fc38) (build 17.0.8+7) OpenJDK 64-Bit Server VM (Red_Hat-17.0.8.0.7-1.fc38) (build 17.0.8+7, mixed mode, sharing)

### Quarkus version or git rev

3.5.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

 Welcome to Gradle 8.3!  Here are the highlights of this release:  - Faster Java compilation  - Reduced memory usage  - Support for running on Java 20  For more details see https://docs.gradle.org/8.3/release-notes.html   ------------------------------------------------------------ Gradle 8.3 ------------------------------------------------------------  Build time:   2023-08-17 07:06:47 UTC Revision:     8afbf24b469158b714b36e84c6f4d4976c86fcd5  Kotlin:       1.9.0 Groovy:       3.0.17 Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023 JVM:          17.0.8 (Red Hat, Inc. 17.0.8+7) OS:           Linux 6.5.9-200.fc38.x86_64 amd64

### Additional information

_No response_

--- a/extensions/smallrye-reactive-messaging/deployment/src/main/resources/dev-ui/qwc-smallrye-reactive-messaging-channels.js
+++ b/extensions/smallrye-reactive-messaging/deployment/src/main/resources/dev-ui/qwc-smallrye-reactive-messaging-channels.js
@@ -61,7 +61,7 @@ export class QwcSmallryeReactiveMessagingChannels extends LitElement {
                     </vaadin-grid-column>
 
                     <vaadin-grid-column auto-width
-                                        header="Publisher"
+                                        header="Publisher(s)"
                                         ${columnBodyRenderer(this._channelPublisherRenderer, [])}>>
                     </vaadin-grid-column>
 
@@ -95,9 +95,19 @@ export class QwcSmallryeReactiveMessagingChannels extends LitElement {
     }
 
     _channelPublisherRenderer(channel) {
-        const publisher = channel.publisher;
-        if (publisher) {
-            return this._renderComponent(publisher);
+        const publishers = channel.publishers;
+        if (publishers) {
+            if (publishers.length === 1) {
+                return this._renderComponent(publishers[0]);
+            } else if (publishers.length > 1) {
+                return html`
+                  <ul class="smaller">
+                    ${publishers.map(item => html`<li>${this._renderComponent(item)}</li>`)}
+                  </ul>
+                `;
+            } else {
+                return html`<em>No publishers</em>`
+            }
         }
     }
 

--- a/extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/devui/DevReactiveMessagingInfos.java
+++ b/extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/devui/DevReactiveMessagingInfos.java
@@ -34,21 +34,24 @@ public class DevReactiveMessagingInfos {
                         .get();
 
                 // collect all channels
-                Map<String, Component> publishers = new HashMap<>();
+                Map<String, List<Component>> publishers = new HashMap<>();
                 Map<String, List<Component>> consumers = new HashMap<>();
                 Function<String, List<Component>> fun = e -> new ArrayList<>();
 
                 // Unfortunately, there is no easy way to obtain the connectors metadata
                 Connectors connectors = container.instance(Connectors.class).get();
-                publishers.putAll(connectors.outgoingConnectors);
+                for (Entry<String, Component> entry : connectors.outgoingConnectors.entrySet()) {
+                    publishers.computeIfAbsent(entry.getKey(), fun)
+                            .add(entry.getValue());
+                }
                 for (Entry<String, Component> entry : connectors.incomingConnectors.entrySet()) {
                     consumers.computeIfAbsent(entry.getKey(), fun)
                             .add(entry.getValue());
                 }
 
                 for (EmitterConfiguration emitter : context.getEmitterConfigurations()) {
-                    publishers.put(emitter.name(),
-                            new Component(ComponentType.EMITTER,
+                    publishers.computeIfAbsent(emitter.name(), fun)
+                            .add(new Component(ComponentType.EMITTER,
                                     emitter.broadcast() ? "<span class=\"annotation\">&#64;Broadcast</span> "
                                             : "" + asCode(DevConsoleRecorder.EMITTERS.get(emitter.name()))));
                 }
@@ -58,23 +61,27 @@ public class DevReactiveMessagingInfos {
                                     asCode(DevConsoleRecorder.CHANNELS.get(channel.channelName))));
                 }
                 for (MediatorConfiguration mediator : context.getMediatorConfigurations()) {
-                    boolean isProcessor = !mediator.getIncoming().isEmpty() && mediator.getOutgoing() != null;
+                    boolean isProcessor = !mediator.getIncoming().isEmpty() && !mediator.getOutgoings().isEmpty();
                     if (isProcessor) {
-                        publishers.put(mediator.getOutgoing(),
-                                new Component(ComponentType.PROCESSOR, asMethod(mediator.methodAsString())));
+                        for (String outgoing : mediator.getOutgoings()) {
+                            publishers.computeIfAbsent(outgoing, fun)
+                                    .add(new Component(ComponentType.PROCESSOR, asMethod(mediator.methodAsString())));
+                        }
                         for (String incoming : mediator.getIncoming()) {
                             consumers.computeIfAbsent(incoming, fun)
                                     .add(new Component(ComponentType.PROCESSOR,
                                             asMethod(mediator.methodAsString())));
                         }
-                    } else if (mediator.getOutgoing() != null) {
-                        StringBuilder builder = new StringBuilder();
-                        builder.append(asMethod(mediator.methodAsString()));
-                        if (mediator.getBroadcast()) {
-                            builder.append("[broadcast: true]");
+                    } else if (!mediator.getOutgoings().isEmpty()) {
+                        for (String outgoing : mediator.getOutgoings()) {
+                            StringBuilder builder = new StringBuilder();
+                            builder.append(asMethod(mediator.methodAsString()));
+                            if (mediator.getBroadcast()) {
+                                builder.append("[broadcast: true]");
+                            }
+                            publishers.computeIfAbsent(outgoing, fun)
+                                    .add(new Component(ComponentType.PUBLISHER, builder.toString()));
                         }
-                        publishers.put(mediator.getOutgoing(),
-                                new Component(ComponentType.PUBLISHER, builder.toString()));
                     } else if (!mediator.getIncoming().isEmpty()) {
                         for (String incoming : mediator.getIncoming()) {
                             consumers.computeIfAbsent(incoming, fun)
@@ -113,12 +120,12 @@ public class DevReactiveMessagingInfos {
     public static class DevChannelInfo implements Comparable<DevChannelInfo> {
 
         private final String name;
-        private final Component publisher;
+        private final List<Component> publishers;
         private final List<Component> consumers;
 
-        public DevChannelInfo(String name, Component publisher, List<Component> consumers) {
+        public DevChannelInfo(String name, List<Component> publishers, List<Component> consumers) {
             this.name = name;
-            this.publisher = publisher;
+            this.publishers = publishers != null ? publishers : Collections.emptyList();
             this.consumers = consumers != null ? consumers : Collections.emptyList();
         }
 
@@ -126,8 +133,8 @@ public class DevReactiveMessagingInfos {
             return name;
         }
 
-        public Component getPublisher() {
-            return publisher;
+        public List<Component> getPublishers() {
+            return publishers;
         }
 
         public List<Component> getConsumers() {
@@ -136,27 +143,17 @@ public class DevReactiveMessagingInfos {
 
         @Override
         public int compareTo(DevChannelInfo other) {
-            if (publisher != other.publisher) {
-                if (other.publisher == null) {
-                    return -1;
-                }
-                if (publisher == null) {
-                    return 1;
-                }
-                // publisher connectors first
-                if (publisher.type != other.publisher.type) {
-                    return publisher.type == ComponentType.CONNECTOR ? -1 : 1;
-                }
+            // publisher connectors last
+            long publisherConnectors = publishers.stream().filter(Component::isConnector).count();
+            long otherPublisherConnectors = other.publishers.stream().filter(Component::isConnector).count();
+            if (publisherConnectors != otherPublisherConnectors) {
+                return Long.compare(otherPublisherConnectors, publisherConnectors);
             }
             // consumer connectors last
             long consumerConnectors = consumers.stream().filter(Component::isConnector).count();
             long otherConsumersConnectors = other.consumers.stream().filter(Component::isConnector).count();
             if (consumerConnectors != otherConsumersConnectors) {
                 return Long.compare(otherConsumersConnectors, consumerConnectors);
-            }
-            if (publisher != other.publisher && publisher.type == ComponentType.CONNECTOR
-                    && other.publisher.type != ComponentType.CONNECTOR) {
-                return 1;
             }
             // alphabetically
             return name.compareTo(other.name);

--- a/extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/devui/ReactiveMessagingJsonRpcService.java
+++ b/extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/devui/ReactiveMessagingJsonRpcService.java
@@ -24,7 +24,7 @@ public class ReactiveMessagingJsonRpcService {
     private JsonObject toJson(DevReactiveMessagingInfos.DevChannelInfo channel) {
         JsonObject json = new JsonObject();
         json.put("name", channel.getName());
-        json.put("publisher", toJson(channel.getPublisher()));
+        json.put("publishers", toJson(channel.getPublishers()));
         json.put("consumers", toJson(channel.getConsumers()));
         return json;
     }

--- a/integration-tests/devmode/src/test/java/io/quarkus/test/devui/DevUIReactiveMessagingJsonRPCTest.java
+++ b/integration-tests/devmode/src/test/java/io/quarkus/test/devui/DevUIReactiveMessagingJsonRPCTest.java
@@ -43,9 +43,9 @@ public class DevUIReactiveMessagingJsonRPCTest extends DevUIJsonRPCTest {
                 consumerExists = typeAndDescriptionExist(consumers, "CHANNEL",
                         "<code>io.quarkus.test.devui.MyProcessor#channel</code>");
             }
-            JsonNode publisher = channel.get("publisher");
-            if (publisher != null) {
-                publisherExists = typeAndDescriptionExist(publisher, "PROCESSOR",
+            JsonNode publishers = channel.get("publishers");
+            if (publishers != null) {
+                publisherExists = typeAndDescriptionExist(publishers, "PROCESSOR",
                         "<code>io.quarkus.test.devui.MyProcessor#process()</code>");
             }
         }

